src/jwtserviceauth.test.js

```diff
@@ -0,0 +1,311 @@
+    listenPromise.then(result => {
+      baseUrl = `http://localhost:${result.port}`
+      httpRequestHandlerR2 = (method, url, headers, body) => {
+        // Overwrite to point to test server
+        let parsedUrl = curl.parse(url)
+        url = `${baseUrl}${parsedUrl.path}`
+        // Do http request
+        let r2Request = r2[method.toLowerCase()](url, {
+          headers,
+          body
+        }).response
+        return r2Request.then(httpResponse => {
+          return httpResponse.arrayBuffer().then(data => {
+            return {
+              statusCode: httpResponse.status,
+              data: data,
+              headers: httpResponse.headers
+            }
+          })
+        })
+      }
+      console.log(`Listining on ${result.hostname}:${result.port}`)
+      done()
+    })
+  })
+
+  after(() => {
+    httpServer.close()
+  })
+
+  describe('httpRequest', () => {
+    it('should return 404', () => {
+      let response = httpRequest('GET', baseUrl)
+      return expect(response, 'to be fulfilled with value satisfying', {
+        statusCode: 404
+      })
+    })
+    it('should return too large', () => {
+      let response = httpRequest(
+        'GET',
+        `${baseUrl}/large_response`,
+        null,
+        null,
+        { maxResponseSize: 512 }
+      )
+      return expect(
+        response,
+        'to be rejected with error satisfying',
+        new HttpRequestError('Response too lange')
+      )
+    })
+    it('should return POST data', () => {
+      let response = httpRequest('POST', `${baseUrl}/echo`, null, 'Hello')
+      return expect(response, 'to be fulfilled with value satisfying', {
+        statusCode: 200
+      })
+    })
+    it('should timeout', () => {
+      let response = httpRequest('GET', `${baseUrl}/timeout`, null, null, {
+        timeout: 1
+      })
+      return expect(
+        response,
+        'to be rejected with error satisfying',
+        new HttpRequestError('Timeout')
+      )
+    })
+  })
+
+  describe('JwtServiceAuthError', () => {
+    it('innerError should be null', () => {
+      let error = new JwtServiceAuthError('')
+      expect(error.innerError, 'to be null')
+    })
+  })
+
+  describe('getGoogleAccessToken', () => {
+    it('should succeed with ok token', () => {
+      let jwtServiceAuth = new JwtServiceAuth(httpRequestHandlerR2)
+      let accessTokenPromise = jwtServiceAuth.getGoogleAccessToken(
+        JSON.stringify(googleKeyFileData)
+      )
+      return expect(
+        accessTokenPromise,
+        'to be fulfilled with value satisfying',
+        {
+          accessToken: 'ok',
+          expiresIn: 3600
+        }
+      )
+    })
+    it('should fail', () => {
+      let jwtServiceAuth = new JwtServiceAuth(httpRequestHandlerR2)
+      return jwtServiceAuth
+        .getGoogleAccessToken(JSON.stringify(googleKeyFileData), 3600, [])
+        .then(accessToken => {
+          return new Error('Got back accessToken when errors was expected')
```

Comment:
Should we ever return an error object? Shouldn't it always be thrown, so it goes into catch etc?

Also consider using expect.fail in this case?

Reply:
We can't throw inside a promise so we need to return the error instead, mocha will handle this for us. I'm supporting node 6.x so I can't use async/await